81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,230 @@
+
+Workflow routes defines the transformation paths from input schemas to output schemas using specific workflows. Each workflow route specifies which workflow to apply for a given input schema and what output schema to produce. 
+
+Workflow routes also define a a graph structure where schemas are nodes and workflow routes are directed edges. The source nodes of this graph are the “original schemas” aka EM models defined as schemapacks. The graph must not contain any cycles, i.e. it should be a directed acyclic graph (DAG). Additionally, the graph must not contain any "diamonds", i.e. there must be at most one directed path between any two schemas in the graph.
+
+This graph structure is used to validate the transformation configuration and to determine the order in which schemas need to be processed during service startup and when processing incoming AnnotatedEMPacks.
+
+#### Transformation Configuration Validation
+
+Transformation configuration validation includes:
+
+1. Check the graph follows a tree-like structure, i.e. it conforms to the unique path requirement mentioned above.
+2. Calculate the topological order of the nodes in the graph using an algorithm similar to Kahn's algorithm, so that the schemas are ordered in a way that if we process the schemas in that order, the previous schemas have already been processed.
+3. Update the `order` field of each schema in the `Schemas` according to the calculated topological order. 
+4. Check the workflows and the original schemas referred by the workflow routes exist in their corresponding collections.
+
+#### User Journeys: Schema Derivation (Manual Trigger)
+
+When manually triggered, the service derives the output schemas for all workflow routes with empty output schemas. 
+
+1. Validate the transformation configuration of the workflow routes.
+2. Traverse the schemas in the transformation graph starting with the original schema, in topological order. For every route execute the following:
+   1. Retrieve the workflow of the corresponding `workflow_id` from the `Workflows`
+   2. Retrieve the schema corresponding to the `input_schema_id` from the `WorkflowRoutes`.
+   3. Call metldata to execute the workflow on the input schema to compute the derived schema.
+   4. Record the derived schema in the `Schemas` and save its id to the `output_schema_id` of the WorkflowRoute.
+3. If there are any errors / conflicts, the operation shall fail and report the error.
+
+
+#### User Journeys: Service Consumer Receives An Original AnnotatedEMPack
+
+The transformation operation is triggered when the service consumer receives "original annotated Em Pack" or a “re-transform metadata” event.
+
+1. Fetch the datapack ids and corresponding schema ids of all datapacks in the collection where the original_id field matches our original datapack id. Create a mapping of these schema ids to their datapack ids that we call “dirty map”, because all of these datapacks need to be either re-created or deleted.
```

Comment:
```suggestion
1. Fetch the datapack ids and corresponding schema ids of all datapacks in the AnnotatedEMPacks collection where the `original_id` field matches our original datapack id. Create a mapping of these schema ids to their datapack ids that we call “dirty map”, because all of these datapacks need to be either re-created or deleted.
```